go/vt/vttablet/tabletmanager/vreplication/external_connector_test.go

```diff
@@ -0,0 +1,157 @@
+		fmt.Sprintf("create table %s.tab2(id int, val varbinary(128), primary key(id))", vrepldb),
+		"insert into tab2 values(1, 'a'), (2, 'b')",
+
+		"create table tab3(id int, val varbinary(128), primary key(id))",
+		fmt.Sprintf("create table %s.tab3(id int, val varbinary(128), primary key(id))", vrepldb),
+		"insert into tab3 values(1, 'a'), (2, 'b')",
+	})
+	defer execStatements(t, []string{
+		"drop table tab1",
+		fmt.Sprintf("drop table %s.tab1", vrepldb),
+		"drop table tab2",
+		fmt.Sprintf("drop table %s.tab2", vrepldb),
+		"drop table tab3",
+		fmt.Sprintf("drop table %s.tab3", vrepldb),
+	})
+
+	filter1 := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match:  "tab1",
+			Filter: "",
+		}},
+	}
+	bls1 := &binlogdatapb.BinlogSource{
+		ExternalMysql: "exta",
+		Filter:        filter1,
+	}
+	cancel1 := startExternalVReplication(t, bls1)
+
+	expectDBClientQueries(t, []string{
+		"begin",
+		"insert into tab1(id,val) values (1,'a'), (2,'b')",
+		"/update _vt.copy_state",
+		"commit",
+		"/delete from _vt.copy_state",
+		"/update _vt.vreplication set state='Running'",
+	})
+	execStatements(t, []string{"insert into tab1 values(3, 'c')"})
+	expectDBClientQueries(t, []string{
+		"begin",
+		"insert into tab1(id,val) values (3,'c')",
+		"/update _vt.vreplication set pos=",
+		"commit",
+	})
+	// Cancel immediately so we don't deal with spurious updates.
+	cancel1()
+
+	// Check that only one connector was created.
+	assert.Equal(t, 1, len(playerEngine.ec.connectors))
+
+	filter2 := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match:  "tab2",
+			Filter: "",
+		}},
+	}
+	bls2 := &binlogdatapb.BinlogSource{
+		ExternalMysql: "exta",
+		Filter:        filter2,
+	}
+	cancel2 := startExternalVReplication(t, bls2)
+
+	expectDBClientQueries(t, []string{
+		"begin",
+		"insert into tab2(id,val) values (1,'a'), (2,'b')",
+		"/update _vt.copy_state",
+		"commit",
+		"/delete from _vt.copy_state",
+		"/update _vt.vreplication set state='Running'",
+	})
+	cancel2()
+
+	// Check that only one connector was created.
+	assert.Equal(t, 1, len(playerEngine.ec.connectors))
+
+	filter3 := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match:  "tab3",
+			Filter: "",
+		}},
+	}
+	bls3 := &binlogdatapb.BinlogSource{
+		ExternalMysql: "extb",
+		Filter:        filter3,
+	}
+	cancel3 := startExternalVReplication(t, bls3)
+
+	expectDBClientQueries(t, []string{
+		"begin",
+		"insert into tab3(id,val) values (1,'a'), (2,'b')",
+		"/update _vt.copy_state",
+		"commit",
+		"/delete from _vt.copy_state",
+		"/update _vt.vreplication set state='Running'",
+	})
+	cancel3()
+
+	// Check that there now two connectors.
+	assert.Equal(t, 2, len(playerEngine.ec.connectors))
+}
+
+func startExternalVReplication(t *testing.T, bls *binlogdatapb.BinlogSource) (cancelr func()) {
+	query := binlogplayer.CreateVReplication("test", bls, "", 9223372036854775807, 9223372036854775807, 0, vrepldb)
```

Comment:
It seems that we are only testing `vcopy` in this test. I think we should also test a externalVReplication with a starting position provided. That will ensure that we go through the main methods of the interface (VStream and VStreamRows). I think this got lost when deleting the old tests.